### PR TITLE
UI lib refactoring

### PR DESCRIPTION
**Pull request by frzyc**
## Describe your changes

 - Some UI refactoring to make it easier to leverage some of the existing components in `frontend` to be used by other apps/libs
 - modularize `Theme`

## Testing/validation

Run `yarn run mini-ci`

## Checklist before requesting a review (leave this PR as draft if any part of this list is not done.)

- [x] I have commented my code, in hard-to understand areas.
- [x] I have made corresponding changes to README or wiki.
- [x] For front-end changes, I have updated the corresponding English translations.
- [x] Ran `yarn run mini-ci` locally to validate format + lint.
- [x] If there were format issues, I ran `nx format write` to resolve them automatically.

### CODE REVIEW

File: libs/gi-ui/project.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "gi-ui",
+  "$schema": "../../node_modules/nx/schemas/project-schema.json",
+  "sourceRoot": "libs/gi-ui/src",
```
Comment (nguyentvan7) [inline libs/gi-ui/project.json:4]: Please add this to ignore-paths in https://github.com/frzyc/genshin-optimizer/blob/master/.github/workflows/deploy-sro-pr.yml

**Review by nguyentvan7 — CHANGES_REQUESTED** [review]

File: apps/gi-frontend/tsconfig.json [hunk 2]
```diff
@@ -6,7 +6,8 @@
     "types": ["vite/client", "vitest"],
     "module": "ESNext",
     "resolveJsonModule": true,
-    "allowSyntheticDefaultImports": true
+    "allowSyntheticDefaultImports": true,
+    "exactOptionalPropertyTypes": false
```
Comment (nguyentvan7) [inline apps/gi-frontend/tsconfig.json:10]: Why?

File: tsconfig.base.json [hunk 3]
```diff
@@ -35,12 +35,19 @@
...
       ],
       "@genshin-optimizer/gi-formula": ["libs/gi-formula/src/index.ts"],
+      "@genshin-optimizer/gi-frontend-gql": [
+        "libs/gi-frontend-gql/src/index.ts"
+      ],
       "@genshin-optimizer/gi-good": ["libs/gi-good/src/index.ts"],
       "@genshin-optimizer/gi-localization": [
         "libs/gi-localization/src/index.ts"
       ],
+      "@genshin-optimizer/gi-prisma-schema": [
```
Comment (nguyentvan7) [inline tsconfig.base.json:45]: what is this and gi-frontend-gql? They don't seem to be in this PR

File: libs/ui-common/tsconfig.json [hunk 4]
```diff
@@ -2,7 +2,8 @@
   "compilerOptions": {
     "jsx": "react-jsx",
     "esModuleInterop": false,
-    "jsxImportSource": "@emotion/react"
+    "jsxImportSource": "@emotion/react",
+    "exactOptionalPropertyTypes": false
```
Comment (nguyentvan7) [inline libs/ui-common/tsconfig.json:6]: why pt2

File: libs/gi-ui/src/App.scss [hunk 5]
```diff
@@ -0,0 +1,16 @@
+/* For Rarity background*/
```
Comment (nguyentvan7) [inline libs/gi-ui/src/App.scss:1]: Should we delete the frontend/app.scss and move it to use this? or is that already done if we use the gi-ui components? I just don't want to have duplicated css where if we change it, we need to change it in 2 places

File: tsconfig.base.json (hunk 3 above)
Comment (frzyc) [reply tsconfig.base.json:45]: oh weird, I was sure i removed these

File: apps/gi-frontend/tsconfig.json (hunk 2 above)
Comment (frzyc) [reply apps/gi-frontend/tsconfig.json:10]: with `exactOptionalPropertyTypes` on true, a lot of properties needs to be super defined:
Like you cant do 
```
type Prop={
  something?:string
}
function Com({something}:Prop){...}

// will typeerror out saying string|undefined can't be used
<Com something={cond?"string":undefined} />
```

you will have to explicitly type it to be 
```
type Prop ={
  something?:string|undefined
}
```

File: libs/gi-ui/src/App.scss (hunk 5 above)
Comment (frzyc) [reply libs/gi-ui/src/App.scss:1]: I don't think any of the frontend is using gi-ui component for this reason.
For frontend to use gi-ui, I think we'd need to amend app.scss, and also update `Theme` with the modular theme.

File: apps/gi-frontend/tsconfig.json (hunk 2 above)
Comment (nguyentvan7) [reply apps/gi-frontend/tsconfig.json:10]: If you think that level of safety isn't needed then I guess it's fine to remove

Comment (frzyc) [reply apps/gi-frontend/tsconfig.json:10]: I'll see if i can refactor this option away. I do agree that this is technically more safety, but I remember i struggled a lot with this when i move the components over. (A lot of libs don't have this option enabled when they wrote their typings, which makes "legal" code have red squiggles)

Comment (frzyc) [reply apps/gi-frontend/tsconfig.json:10]: yeah, looks like mui does not support this option. I think any lib with mui components will need this. option

File: libs/gi-ui/src/App.scss (hunk 5 above)
Comment (frzyc) [reply libs/gi-ui/src/App.scss:1]: ... or just change implementation to use the new `content-` palettes

**Review by nguyentvan7 — APPROVED** [review]

**Pull request merged**